src/ai/defaultai.cc

```diff
@@ -269,13 +267,34 @@ void DefaultAI::think() {
 
 	SchedulerTaskId due_task = SchedulerTaskId::kUnset;
 
-	sort_task_pool();
+	// we need to get more one most urgent job
+	sort_task_pool(1);
 
 	const int32_t delay_time = gametime.get() - taskPool.front().due_time.get();
 
-	// Here we decide how many jobs will be run now (none - 5)
-	// in case no job is due now, it can be zero
-	uint32_t jobs_to_run_count = (delay_time < 0) ? 0 : 1;
+	// This portion of code tries to keep the speed of game high, but so that AI's jobs
+	// have reasonably enough time for execution and are not too much delayed
+	if (game().is_auto_speed()) {
+		int32_t speed_diff = 0;
+		if (delay_time > 4000) {
+			speed_diff = -100;
+		} else if (delay_time < 1000) {
+			speed_diff = +100;
+		}
+		if (speed_diff != 0) {
+			if (GameController* const ctrl = game().game_controller()) {
+				if ((ctrl->desired_speed() > 950 && ctrl->desired_speed() < 30000) ||
+				    (ctrl->desired_speed() < 1000 && speed_diff > 0) ||
+				    (ctrl->desired_speed() > 29999 && speed_diff < 0)) {
+					ctrl->set_desired_speed(ctrl->desired_speed() + speed_diff);
+				}
+			}
+		}
+	}
+
+	// Here we decide limit for jobs will be run now
+	// generelly 3, if situation is wrong: 4
+	uint32_t jobs_to_run_count = std::min<uint32_t>((delay_time > 3000) ? 5 : 3, taskPool.size());
```

Comment:
```suggestion
	uint32_t jobs_to_run_count = std::min<uint32_t>(delay_time / 2000 +1, taskPool.size() / 2);
```

Reply:
taskPool.size() is there only to make sure we do not exceed its size. For case something wrong happen and it is not properly filled. Should not happen, but this will save us from one of two rows checking if jobs_to_run are lesser than the count of our tasks.
What if delay will be 15 seconds, is 8 jobs fine for you?

Reply:
in theory yes of course trying to get the delay removed by computing more tasks would be good. But thewre is a frontier if the AI uses nearly all of the processing time it creates delay itself so we might run into a stall if we don't limit this to a certain value (was 4 before). i am fully aware this would limit the jobs to 7. but this should be fine as we probably won't be able to manage the situation if delay still grows although this many jobs get run

Reply:
I am lost. I suggest to limit jobs to run to 3 or 5, and this is too much for you, but 7 is OK?

Reply:
This is not a contradiction in my eyes. I just want to ensure that the cpu time we require is distributed over time as equally as possible. so if everything is normal (let's say gamespeed 1x to 3x) and our delay is less then 2000ms this is best achieved by only running 1 job per think() cycle. If we really are getting late we need to to do something so in this case the requirement to distribute CPU times can't apply anymore but we need to use as much of it (together with the logic of the game) as we can to get our jobs done. So increasing it up to 7 should be a last ressort chance. 

the problem with starting with 3 would be that if the first 3 tasks will be due (i.e. duetime < gametime) they will be executed in one cycle. followed by perhaps 2 idle cycles in an ormal game. this would cause the game to stutter. so the main thing is to staret from 1 and only increase if needed.

Reply:
Even now, more than 1 task is run every think() - I guarantee. So I do not propose anything worse that it is now.

What about my suggestion to execute a job only if:
delay_time > position_in_queue * 300 ms?

Reply:
the second job is run only if the first task has a delay of more then 2000 ms according to line
https://github.com/widelands/widelands/blob/b1df95a2a3aa80d2dd50aeccc43630bf9bb1eddf/src/ai/defaultai.cc#L301

so this can happen from time to time But I think normally only 1 task is run. I will try whether you might be right with some debug code though.

Reply:
So you need delay 4,5 sec to run 3 jobs and 8 sec to run 4.... bit limiting, but do you tests...

Reply:
first test: 2 AI on crater. Search for jobs in the log output and see the latest commit in my branch for the loggging code.
result: even with gamespeed 8x mostly 1 job is run per cycle.  only at 16x sometimes a 3rd job is necessary.

Reply:
what about bigger map with 4 AI players with auto_speed?

Reply:
which branch? Your fix_auto_speed?

Reply:
OK, looks good indeed. I will compile too and test myself

Reply:
I just tested for a short time maybe we need to look at a long game with bigger map

Reply:
test runs :)

Reply:
AI just does not grow, so no point to wait long... It looks OK, we can stay with this. However I would comment your branch mostly to get rid of some asserts.
we can always return and finetune this later

Reply:
Waht do you mean by AI just don't grow?

Reply:
I hoped that after 2 hours of gametime it would be bigger and meat each other, have bigger territory

Reply:
But this is different topic

Reply:
It does exactly this in my setup. 
Which map did you use? could you provide me a replay? 
Currently a 8 AI match on Lesser ring is running at my side,

On crater for example usually the match between 2 AI is over after approx 4hours gametimeor earlier

Reply:
I tried multiple games. Never mind now...
What I noticed it runs much slower than auto_speed games used to run. But this is matter of tuning some values in the logic, I can experiment when training the AI